vm/rtti.h

```diff
@@ -0,0 +1,119 @@
+// vim: set sts=2 ts=8 sw=2 tw=99 et:
+//
+// Copyright (C) 2004-2021 AlliedModers LLC
+//
+// This file is part of SourcePawn. SourcePawn is licensed under the GNU
+// General Public License, version 3.0 (GPL). If a copy of the GPL was not
+// provided with this file, you can obtain it here:
+//   http://www.gnu.org/licenses/gpl.html
+//
+#ifndef _include_sourcepawn_rtti_h_
+#define _include_sourcepawn_rtti_h_
+
+#include <memory>
+#include <vector>
+
+#include <sp_vm_types.h>
+#include <amtl/am-hashmap.h>
+
+namespace sp {
+
+  class Rtti;
+  class SmxV1Image;
+
+  class RttiData {
```

Comment:
It doesn't look like this file is used anywhere in the tree yet? Does it need to be in this changeset?

Reply:
It's used to validate the type information in all the other rtti sections, but the `Rtti` class isn't used yet indeed. I could strip out the decoding functionality and only leave the validation for now, if that seems better.

Reply:
Ack, I missed where it's #included. It's fine for now, though the body should be dedented (namespaces don't increase the indent level).

It'd be better if it was clearly marked as something used for debugging purposes though. Slapping a "debug" namespace around it might be best. The reason is that, I haven't fully mapped out where RTTI is going. We don't currently use it for anything, and it's not clear it would exist in a source-only environment. And if it does, `Rtti` almost certainly won't work for actual runtime stuff.

I don't want to hold up your reviews trying to refactor something for a completely unknown use case, but I also don't want to have to refactor a bunch of dependent code later too.

So keeping it inside a namespace just for its usecase for now seems like the best path forward.

Reply:
Understood. I don't want to pin the binary layout of the rtti/debug info, but that's a necessary evil if we want some state introspection for dynamic debugging (which I really want :D). Since you didn't use the info yet, I didn't push eagerly for these changes to be merged and it always sounded like you're about to use them in some way which could mean changing the layout again.

Can't talk about switching the architecture for a source-only runtime, since that's all on your agenda in the upcoming decades. We can always drop support for debugging again, if it's too much of a hassle.

The Rtti classes live in the `sp::debug` namespace now and I'll add all the other code regarding debugging support in there as well.